File: e2e-tests/cmd/go/main.go
// SPDX-License-Identifier: MPL-2.0
/*
 * libpathrs: safe path resolution on Linux
 * Copyright (C) 2019-2025 SUSE LLC
 * Copyright (C) 2026 Aleksa Sarai <cyphar@cyphar.com>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package main

import (
	"context"
	"errors"
	"fmt"
	"os"
	"strconv"
	"syscall"

	"github.com/urfave/cli/v3"
	"golang.org/x/sys/unix"
)

func Main(args []string) error {
	if dumpableStr := os.Getenv("PR_SET_DUMPABLE"); dumpableStr != "" {
		dumpable, err := strconv.ParseUint(dumpableStr, 10, 64)
		if err != nil {
			return fmt.Errorf("invalid PR_SET_DUMPABLE environment variable value: %w", err)
		}
		if err := unix.Prctl(unix.PR_SET_DUMPABLE, uintptr(dumpable), 0, 0, 0); err != nil {
			return fmt.Errorf("failed to set PR_SET_DUMPABLE=%v: %w", dumpable, err)
		}
	}

	cmd := &cli.Command{
		Name:  "pathrs-cmd",
		Usage: "helper binary for testing libpathrs",
		Authors: []any{
			"Aleksa Sarai <cyphar@cyphar.com>",
		},
		Commands: []*cli.Command{
			rootCmd,
			procfsCmd,
		},
	}
	return cmd.Run(context.Background(), args)
}

func main() {
	if err := Main(os.Args); err != nil {
		var errno syscall.Errno
		if errors.As(err, &errno) {
			fmt.Fprintf(os.Stderr, "ERRNO %d (%s)\n", errno, errno)
		}
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

File: e2e-tests/cmd/rust/src/main.rs
// SPDX-License-Identifier: MPL-2.0
/*
 * libpathrs: safe path resolution on Linux
 * Copyright (C) 2019-2025 SUSE LLC
 * Copyright (C) 2026 Aleksa Sarai <cyphar@cyphar.com>
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

use pathrs::error::{Error as PathrsError, ErrorKind as PathrsErrorKind};

use std::process::ExitCode;

use anyhow::{anyhow, Error};
use clap::Command;
use errno::Errno;

mod procfs;
mod root;
mod utils;

fn cli() -> Command {
    Command::new("pathrs-cmd")
        .author("Aleksa Sarai <cyphar@cyphar.com>")
        .subcommand(root::cli())
        .subcommand(procfs::cli())
}

#[test]
fn verify_app() {
    cli().debug_assert();
}

fn handle_error(func: impl FnOnce() -> Result<(), Error>) -> ExitCode {
    if let Err(err) = func() {
        let mut desc = err.to_string();
        for cause in err.chain() {
            if let Some(err) = cause.downcast_ref::<PathrsError>() {
                // This is basically ErrorKind::errno (which isn't exported
                // currently). This is necessary in order to emulate the
                // behaviour of the binding test programs, as they see the
                // converted errnos as well as the legitimate OsError ones.
                let errno = match err.kind() {
                    PathrsErrorKind::NotImplemented => Some(libc::ENOSYS),
                    PathrsErrorKind::InvalidArgument => Some(libc::EINVAL),
                    PathrsErrorKind::OsError(errno) => errno,
                    PathrsErrorKind::SafetyViolation => Some(libc::EXDEV),
                    _ => None,
                };
                if let Some(errno) = errno {
                    println!("ERRNO {errno} ({})", Errno(errno));
                }
            }
            // Emulate capi's error formatting.
            desc.push_str(": ");
            desc.push_str(&cause.to_string());
        }
        println!("error: {desc}");
        ExitCode::FAILURE
    } else {
        ExitCode::SUCCESS
    }
}

fn main() -> ExitCode {
    handle_error(|| {
        let mut app = cli();

        match app.get_matches_mut().subcommand() {
            Some(("root", sub_matches)) => root::subcommand(sub_matches),
            Some(("procfs", sub_matches)) => procfs::subcommand(sub_matches),
            Some((subcommand, _)) => {
                // We should never end up here.
                app.print_help()?;
                Err(anyhow!("unknown subcommand '{}'", subcommand))
            }
            None => {
                app.print_help()?;
                Err(anyhow!("no subcommand specified"))
            }
        }
    })
}
